File: tauri/src/sys.rs
use std::{env, fs, path::PathBuf};

use anyhow::Result;

pub fn read_hosts() -> Result<String> {
	let hosts_path = get_hosts_path()?;
	let content = fs::read_to_string(hosts_path)?;

	Ok(content)
}

pub fn write_hosts(mut content: String) -> Result<()> {
	if cfg!(windows) {
		content = content.replace('\n', "\r\n");
	}

	let hosts_path = get_hosts_path()?;
	let tmp_file = env::temp_dir().join("hed2_tmp");
	fs::write(&tmp_file, content)?;
	fs::copy(&tmp_file, hosts_path)?;
	fs::remove_file(&tmp_file)?;

	Ok(())
}

pub fn check_hosts_readonly() -> Result<()> {
	let hosts_path = get_hosts_path()?;
	let permissions = fs::metadata(hosts_path)?.permissions();

	if permissions.readonly() {
		anyhow::bail!(
			"The hosts file is in read-only mode, please disable it manually"
		)
	}

	Ok(())
}

pub fn open_hosts_dir() -> Result<()> {
	let hosts_path = get_hosts_path()?;
	if let Some(dir) = hosts_path.parent() {
		opener::open(dir)?;
	}

	Ok(())
}

#[cfg(all(not(dev), windows))]
fn get_hosts_path() -> Result<PathBuf> {
	let sys_drive = env::var("SYSTEMDRIVE")?;

	Ok(PathBuf::from(format!(
		r"{}\Windows\System32\drivers\etc\hosts",
		sys_drive
	)))
}

#[cfg(all(not(dev), unix))]
fn get_hosts_path() -> Result<PathBuf> {
	Ok(PathBuf::from("/etc/hosts"))
}

#[cfg(dev)]
fn get_hosts_path() -> Result<PathBuf> {
	let mut cwd = env::current_dir()?;

	cwd.push("tmp");
	cwd.push("hosts");

	Ok(cwd)
}

File: electron/src/sys.ts
import fs from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import { IS_DEV } from './consts.js';

export async function readHosts(): Promise<string> {
  const hostsPath = await getHostsPath();
  return fs.readFile(hostsPath, 'utf-8');
}

export async function writeHosts(content: string): Promise<void> {
  const hostsPath = await getHostsPath();
  const tmpFile = path.join(tmpdir(), 'hed2_tmp');
  await fs.writeFile(
    tmpFile,
    process.platform === 'win32' ? content.replaceAll('\n', '\r\n') : content,
  );
  await fs.copyFile(tmpFile, hostsPath);
  await fs.rm(tmpFile);
}

export async function isHostsReadonly(): Promise<boolean> {
  const hostsPath = await getHostsPath();
  try {
    await fs.access(hostsPath, fs.constants.W_OK);
    return true;
  } catch {
    return false;
  }
}

export async function getHostsPath(): Promise<string> {
  if (IS_DEV) {
    return path.join(import.meta.dirname, '../', '../', 'tmp', 'hosts');
  }
  if (process.platform === 'win32') {
    const systemDrive = process.env.SYSTEMDRIVE;
    if (!systemDrive) {
      throw new Error('SYSTEMDRIVE environment variable is not set');
    }
    return path.join(
      systemDrive,
      'Windows',
      'System32',
      'drivers',
      'etc',
      'hosts',
    );
  }
  return '/etc/hosts';
}
